Fix report

mkangia opened the pull request:
https://manage.dimagi.com/default.asp?277767#BugEvent.1508932

From that made the following updates to keep the report simple but functional for now
1. Allow sorting only by product
2. Changes to sort by product

its almost ready except a bug in pagination. best review 🐟 / 🐟

mkangia committed:
make only product column sortable

mkangia committed:
sort by product name and respect pagination

mkangia committed:
dont/cant cache now

mkangia committed:
return just unique product ids from ledgers

mkangia committed:
remove unused import

mkangia committed:
better import structure

mkangia committed:
filter for products with ledgers always

mkangia committed:
extract properties to avoid multiple fetch

mkangia committed:
add some documentation

mkangia committed:
fix order

mkangia commented:
@proteusvacuum this is ready..

mkangia committed:
lint

esoergel commented on corehq/apps/reports/analytics/esaccessors.py:
```diff
@@ -578,11 +578,15 @@ def get_wrapped_ledger_values(domain, case_ids, section_id, entry_ids=None, pagi
     if entry_ids:
         query = query.entry(entry_ids)
-    return query.count()
+    return set(query.values_list('entry_id', flat=True))
```
Does this report only show products that have entries associated with them?  I'd expect it to include all products and just show `0` or `---` for the ones without data.  I understand changing that may be out of scope if you're not confident.

mkangia replied:
I kept things according to the original behaviour even though the report earlier was broken.
Plus including blanks for missing products would need to relook at the [logic](https://github.com/dimagi/commcare-hq/pull/21101/files#diff-007cbd344b264b88f44398660b42183aR216) where data is being built and iterate differently.
So keeping this.

esoergel commented on corehq/apps/reports/commtrack/standard.py:
```diff
@@ -109,19 +109,19 @@ def shared_pagination_GET_params(self):
 
-    @memoized
     @property
+    @memoized
```
You can also use `django.utils.functional.cached_property` for these - I also always mess up the order :tongue:

mkangia replied:
*legends* :)

esoergel commented on corehq/apps/reports/commtrack/standard.py:
```diff
@@ -151,16 +179,37 @@ def headers(self):
+        # product ids to filter
+        # -> all that have ledgers and
+        # -> if specific program products ids from filter
+        product_ids_to_filter = products_with_ledgers(self.domain, self._sp_ids, STOCK_SECTION_TYPE, None)
```
This is being called twice during report generation - here, and again to get the count.  You could make it a memoized class property instead, and for count, just call `len()` on it.

esoergel commented on corehq/apps/reports/commtrack/standard.py:
```diff
@@ -151,16 +179,37 @@ def headers(self):
+                            if _product_name_map[0] in product_ids_to_filter]
+        # product ids according to pagination
+        return [product_id for product_id, product_name in
+                products_to_show[self.pagination.start:][:self.pagination.count]]
```
nit: I think you could combine these last two expressions:
```python
return [
    product_id for product_id, product_name in products_to_show
    if product_id in product_ids_to_filter
][self.pagination.start:][:self.pagination.count]
```

mkangia replied:
Beauty!

esoergel commented on corehq/apps/reports/commtrack/util.py:
```diff
@@ -98,8 +98,11 @@ def filter_relevant(queryset):
+    products = SQLProduct.objects.filter(domain=domain)
+    if product_ids:
+        products = products.filter(product_id__in=product_ids)
+    return dict(products.values_list('product_id', 'name'))
```
Looks like pagination wouldn't be possible, since this includes only products with ledgers, but you could still move filtering by program into the db call.

mkangia replied:
making the change but dont see an advantage..
One thing i like about the current approach is that it finds product ids and then uses only that for both ES and SQL whereas this divides program id for SQL and product ids for ES. Not sure if that sounds as an advantage but i feel it keeps things simple

mkangia replied:
Plus we still need to find product ids for ES. Open to understanding if there is an advantage with the change.

esoergel commented on corehq/apps/reports/commtrack/standard.py:
```diff
@@ -151,16 +179,37 @@ def headers(self):
+        # -> if specific program products ids from filter
+        product_ids_to_filter = products_with_ledgers(self.domain, self._sp_ids, STOCK_SECTION_TYPE, None)
+        if self._program_product_ids:
+            product_ids_to_filter = product_ids_to_filter & set(self._program_product_ids)
```
I think this is redundant - shouldn't `sorted_product_name_map` already be filtered to match the program?

mkangia replied:
nice catch.. thanks..

esoergel reviewed (approved):
The current behavior of including only products with ledgers makes things more complicated than they would otherwise be, and I think is actually worse from a usability perspective.  You could reach out to product and see if they agree, and change this to go over all selected products.

Otherwise I had some comments, but I don't think there were any blockers.

mkangia committed:
review fixes 1

- remove products_with_ledgers_count
- product name map already considers program product ids to no need to check for it again.

mkangia committed:
use program id for filtering from SQL

mkangia committed:
simplify

mkangia committed:
use cached_property instead

mkangia committed:
fix error calling a set returned
